### ISSUE TITLE
[Bug]: URDF vs XML

### ISSUE THREAD

**Issue by ax-anoop** [Genesis-Embodied-AI/Genesis#1407]
### Bug Description

Hey I'm comparing the URDF loading vs XML (for unitree G1) 

<img width="3703" height="1076" alt="Image" src="https://github.com/user-attachments/assets/96c84e13-f742-4d55-bb59-ed66718b3c0b" />

I noticed that with latest update there's bene some improvements in correct parsing and matching the different loading processes, but there still seems to be some gaps. 

[g1_files.zip](https://github.com/user-attachments/files/21206528/g1_files.zip)

- Link data has been unified recently from a few months ago when I pulled (AWESOME) 
- Force limits different (but because xml does not have it, and urdf has "effort") 
- Kp / Kv defaults different, I'm not sure why this is the case ? Neither have these backed into the data files 
**- Damping, URDF data not being loaded** 
- Inverse weight is super different BUT once armature is matched manually, its roughly the same. 

<img width="3703" height="1076" alt="Image" src="https://github.com/user-attachments/assets/8c741882-6bac-4f56-afa2-a9ec5911bfd6" />

URDF Data
  <joint name="left_knee_joint" type="revolute">
    <parent link="left_knee_motor"/>
    <child link="left_knee_link"/>
    <origin xyz="0 -0.043 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="-1" upper="2.8798" effort="139" velocity="20"/>
    <dynamics damping="0.05" friction="0.2"/>
  </joint>

So the main issue I am seeing right now is damping not being loaded, but great job on the recent improvements guys. 

To reproduce:

```
import argparse
import time

import genesis as gs
import numpy as np
import torch

DEVICE = "cpu"

JOINT_PROPERTIES = {
    'left_hip_pitch_joint':   {'torque': 88,  'default': -0.20, 'pgain': 200.0, 'dgain': 5.0},
    'left_hip_roll_joint':    {'torque': 88,  'default': 0.0,   'pgain': 150.0, 'dgain': 5.0},
    'left_hip_yaw_joint':     {'torque': 88,  'default': 0.0,   'pgain': 150.0, 'dgain': 5.0},
    'left_knee_joint':        {'torque': 139, 'default': 0.42,  'pgain': 200.0, 'dgain': 5.0},
    'left_ankle_pitch_joint': {'torque': 50,  'default': -0.23, 'pgain': 20.0,  'dgain': 2.0},
    'left_ankle_roll_joint':  {'torque': 50,  'default': 0.0,   'pgain': 20.0,  'dgain': 2.0},
    'right_hip_pitch_joint':   {'torque': 88,  'default': -0.20, 'pgain': 200.0, 'dgain': 5.0},
    'right_hip_roll_joint':    {'torque': 88,  'default': 0.0,   'pgain': 150.0, 'dgain': 5.0},
    'right_hip_yaw_joint':     {'torque': 88,  'default': 0.0,   'pgain': 150.0, 'dgain': 5.0},
    'right_knee_joint':        {'torque': 139, 'default': 0.42,  'pgain': 200.0, 'dgain': 5.0},
    'right_ankle_pitch_joint': {'torque': 50,  'default': -0.23, 'pgain': 20.0,  'dgain': 2.0},
    'right_ankle_roll_joint':  {'torque': 50,  'default': 0.0,   'pgain': 20.0,  'dgain': 2.0},
}

class Robot:
    def __init__(self, xml_location, render=True, n_envs=1, simple_view=False):
        """
        Initializes the robot simulation.

        Args:
            xml_location (str): Path to the robot's MJCF or URDF file.
            render (bool): If True, enables the simulation viewer.
            n_envs (int): Number of parallel simulation environments.
            simple_view (bool): If True, only prints info for specific components.
        """
        self.xml_location = xml_location
        self.n_envs = n_envs
        self.render = render
        self.simple_view = simple_view

        gs.init(backend=gs.cpu, logging_level="error")
        self._initialize_simulation()
        self.analyze_robot()
        self._initialize_robot_properties()

    def _initialize_simulation(self):
        """
        Initializes the Genesis scene, including the robot and ground plane.
        """
        self.scene = gs.Scene(
            show_viewer=self.render,
            sim_options=gs.options.SimOptions(dt=0.02, gravity=(0, 0, -9.81), substeps=4),
        )
        self.plane = self.scene.add_entity(gs.morphs.Plane())
        if self.xml_location.endswith('.xml'):
            self.robot = self.scene.add_entity(gs.morphs.MJCF(file=self.xml_location))
        elif self.xml_location.endswith('.urdf'):
            self.robot = self.scene.add_entity(
                gs.morphs.URDF(
                    file=self.xml_location,
                    pos=(0, 0, 0.75),
                    euler=(0, 0, 0),
                    scale=1.0,
                    fixed=False,
                    default_armature=0.01
                ),
            )
        else:
            raise ValueError(f"Unsupported file format: {self.xml_location}. Only .xml or .urdf files are supported.")
        self.scene.build(n_envs=self.n_envs, env_spacing=(2.0, 2.0))

    def analyze_robot(self):
        """
        Analyzes the robot's properties.
        """
        self.robot_mass = sum(link.inertial_mass for link in self.robot.links)
        print(f"Calculated total robot mass: {self.robot_mass:.2f} kg")

        print("\n--- Link Inertial Properties ---")
        for link in self.robot.links:
            if self.simple_view and link.name != "left_knee_link":
                # print(f"Skipping link: {link.name}")
                continue
            print(f"\nLink: {link.name}")
            print(f"  Mass: {link.inertial_mass}")
            print(f"  Inertia tensor:\n{link.inertial_i}")
            print(f"  COM position: {link.inertial_pos}")
            print(f"  COM orientation (quat): {link.inertial_quat}")
        print("---------------------------------\n")

    def _initialize_robot_properties(self):
        """
        Sets up the robot's joint controllers, gains, and default positions.
        """
        valid_joints = {}
        for j in self.robot.joints:
            if isinstance(j.dof_idx_local, int):
                valid_joints[j.name] = j.dof_idx_local
        
        self.jnt_names = list(valid_joints.keys())
        dofs_idx_list = list(valid_joints.values())
        
        self.dofs_idx = torch.tensor(dofs_idx_list, dtype=torch.long, device=DEVICE)
        
        # Fetch all properties at once, assuming they return values for all DOFs
        all_pos_lower, all_pos_upper = self.robot.get_dofs_limit()
        all_force_lower, all_force_upper = self.robot.get_dofs_force_range()
        all_armature = self.robot.get_dofs_armature()
        all_kp = self.robot.get_dofs_kp()
        all_kv = self.robot.get_dofs_kv()
        all_damping = self.robot.get_dofs_damping()
        all_stiffness = self.robot.get_dofs_stiffness()
        all_invweight = self.robot.get_dofs_invweight()

        print("\n--- Joint Properties ---")
        for name in self.jnt_names:
            if self.simple_view and name != "left_knee_joint":
                continue
            joint = self.robot.get_joint(name)
            dof_idx = joint.dof_idx_local
            print(f"\nJoint: {name}")

            print(f"  Position Limits: [{all_pos_lower[dof_idx].item():.4f}, {all_pos_upper[dof_idx].item():.4f}]")
            print(f"  Force Limits: [{all_force_lower[dof_idx].item():.2f}, {all_force_upper[dof_idx].item():.2f}]")
            print(f"  Armature: {all_armature[dof_idx].item():.4f}")
            print(f"  Kp: {all_kp[dof_idx].item():.2f}")
            print(f"  Kv: {all_kv[dof_idx].item():.2f}")
            print(f"  Damping: {all_damping[dof_idx].item():.4f}")
            print(f"  Stiffness: {all_stiffness[dof_idx].item():.4f}")
            print(f"  Inverse Weight: {all_invweight[dof_idx].item():.4f}")
        print("---------------------------------\n")

        kps_list = [JOINT_PROPERTIES[name]['pgain'] for name in self.jnt_names]
        kvs_list = [JOINT_PROPERTIES[name]['dgain'] for name in self.jnt_names]
        t_max_list = [JOINT_PROPERTIES[name]['torque'] for name in self.jnt_names]
        default_angles_list = [JOINT_PROPERTIES[name]['default'] for name in self.jnt_names]

        kps = torch.tensor(kps_list, dtype=torch.float32, device=DEVICE)
        kvs = torch.tensor(kvs_list, dtype=torch.float32, device=DEVICE)
        t_max = torch.tensor(t_max_list, dtype=torch.float32, device=DEVICE)
        t_min = -t_max
        base_angles = torch.tensor(default_angles_list, dtype=torch.float32, device=DEVICE)

        self.robot.set_dofs_kp(kp=kps, dofs_idx_local=self.dofs_idx)
        self.robot.set_dofs_kv(kv=kvs, dofs_idx_local=self.dofs_idx)
        self.robot.set_dofs_force_range(lower=t_min, upper=t_max, dofs_idx_local=self.dofs_idx)

        self.default_angles = base_angles.unsqueeze(0).repeat(self.n_envs, 1)

        try:
            self.base_link_idx = self.robot.get_link("pelvis").idx
        except Exception:
            print("Warning: 'pelvis' link not found. Using link 0 for gravity force.")
            self.base_link_idx = 0

    def reset(self):
        self.scene.reset()
        self.robot.set_dofs_position(self.default_angles, self.dofs_idx)

def parse_args():
    parser = argparse.ArgumentParser(description='Simple G1 Robot Simulation with Genesis')
    parser.add_argument('--no-render', dest='render', action='store_false', help='Disable rendering')
    parser.set_defaults(render=True)
    parser.add_argument('--robot_model', type=str, default='env/teb_no_arms/teb_no_arms.urdf', help='Path to robot model file (.xml or .urdf)')
    parser.add_argument('--simple_view', action='store_true', help='Only print info for specific joint and link')
    args = parser.parse_args()
    return args

def main():
    args = parse_args()

    robot = Robot(
        args.robot_model, 
        render=args.render,
        simple_view=args.simple_view
    )

if __name__ == "__main__":
    main()
```

**Comment by duburcqa** [comment]
Yes, we put a lot of efforts into unifying loading of UDF and MJCF files. There is now a single loader under the hood, but a few fields are still ignored for URDF files, typically damping and armature.

Regarding the difference for invweights, this is expected because it must be computed differently for URDF and MJCF files unless the robot has a fixed base and no armature inertia. There is no way around.

**Comment by ax-anoop** [comment]
Great work :) 

I have sim2real setup for locomotion on the G1, will publish some blogs / videos on it as well. 

Damping was the main field I noticed that is within the urdf, but ignored.

**Comment by duburcqa** [comment]
@ax-anoop Any update on this? I would be interested!

**Comment by ax-anoop** [comment]
I belive this is resolved now & we can close ?

**Comment by duburcqa** [comment]
Great!

**Comment by duburcqa** [comment]
Wait no, damping and armature fields are still ignored in URDF files.

**Comment by duburcqa** [comment]
Added parsing of joint damping and friction in URDF file. It is not possible to add motor armature support because it is not part of URDF format specifications.
